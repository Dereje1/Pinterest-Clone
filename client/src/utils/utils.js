export const shuffleImages = (arr) => {
  const shuffled = [];
  while (arr.length) {
    const randIndex = Math.floor(Math.random() * arr.length);
    const [removed] = arr.splice(randIndex, 1);
    shuffled.push(removed);
  }
  return shuffled;
};

const searchInTags = (tags, search) => {
  for (let i = 0; i < tags.length; i += 1) {
    if (tags[i].tag.toLowerCase().includes(search)) {
      return true;
    }
  }
  return false;
};

export const getFilteredPins = (pinList, search) => {
  if (!search) return pinList;
  const filteredPins = pinList.filter(({ owner, imgDescription, tags }) => {
    const isFoundInDescription = imgDescription.toLowerCase().includes(search);
    const isFoundInOwnerName = owner.name.toLowerCase().includes(search);
    const isFoundInTags = searchInTags(tags, search);
    return isFoundInDescription || isFoundInOwnerName || isFoundInTags;
  });
  return filteredPins;
};

export const getZoomedImageStyle = ({
  naturalWidth: imageWidth,
  naturalHeight: imageHeight,
}) => {
  // dynamically resize image
  let { innerWidth, innerHeight } = window;
  // parameter for innerwidth/height adjustment with mobile consideration
  // top(70) + headingheight(50 / 25) + button height (50 / 25)
  innerHeight = innerHeight < 500 ? innerHeight - 120 : innerHeight - 170;
  // minor x direction adjustment for padding too
  innerWidth -= (innerWidth * 0.02);
  let isNoFit = false;
  const ratio = Math.min(innerWidth / imageWidth, innerHeight / imageHeight);
  const newWidth = imageWidth * ratio;
  let parentWidth = newWidth;
  if (newWidth < 500) {
    if (innerWidth < 500) {
      parentWidth = innerWidth;
    } else {
      parentWidth = 500;
    }
    isNoFit = true;
  }
  return {
    imgWidth: `${newWidth}px`,
    parentWidth,
    titleSize: `${newWidth < 500 ? 1.2 : 2}em`,
    subTitleSize: `${newWidth < 500 ? 0.9 : 1.2}em`,
    dateSize: `${newWidth < 500 ? 0.45 : 0.6}em`,
    pinnersSize: '2em',
    isNoFit,
  };
};

export const getPinners = (savedBy) => (savedBy.length > 3
  ? `${savedBy.slice(0, 3).join(', ')} and ${savedBy.length - 3} others`
  : `${savedBy.join(', ')}`);

export const getFormattedDescription = (imgDescription) => (imgDescription.length > 15 ? `${imgDescription.slice(0, 15)}...` : imgDescription);

// eslint-disable-next-line no-promise-executor-return
export const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export const validateURL = (string) => {
  try {
    const url = new URL(string);
    if (url.protocol === 'data:' || url.protocol === 'https:') return string;
    if (url.protocol === 'http:') {
      // convert to https to avoid mixed content warning in console
      return `${string.split(':')[0]}s:${string.split(':')[1]}`;
    }
  } catch (_) {
    return null;
  }
  return null;
};

export const getModalWidth = () => {
  const { innerWidth } = window;
  if (innerWidth < 500) {
    return innerWidth;
  } if (innerWidth > 1000) {
    return innerWidth / 3;
  }
  return innerWidth / 2;
};

export const isDuplicateError = (allPinLinks, picPreview) => {
  for (let i = 0; i < allPinLinks.length; i += 1) {
    const { imgLink, originalImgLink, cloudFrontLink } = allPinLinks[i];
    if (
      imgLink === picPreview
      || originalImgLink === picPreview
      || cloudFrontLink === picPreview
    ) return true;
  }
  return false;
};

export const encodeImageFileAsURL = (imgFile) => new Promise((resolve, reject) => {
  const reader = new FileReader();
  reader.onloadend = () => {
    resolve(reader.result);
  };
  reader.onerror = (err) => {
    reject(err);
  };
  reader.readAsDataURL(imgFile);
});

export const initialDisplayPerScroll = () => (window.innerWidth > 1440 ? 20 : 10);

export const formatDate = (date) => {
  const [, day, mth, year] = new Date(date).toUTCString().split(' ');
  return `${day} ${mth} ${year}`;
};

export const updatePinList = (oldList, newPin) => {
  const indexOfUpdate = oldList.findIndex((p) => p._id === newPin._id);
  return [
    ...oldList.slice(0, indexOfUpdate),
    newPin,
    ...oldList.slice(indexOfUpdate + 1),
  ];
};
