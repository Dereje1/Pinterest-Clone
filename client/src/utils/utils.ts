import { PinType, tagType } from '../interfaces';

export const shuffleImages = (arr: PinType[]) => {
  const shuffled: PinType[] = [];
  while (arr.length) {
    const randIndex = Math.floor(Math.random() * arr.length);
    const [removed] = arr.splice(randIndex, 1);
    shuffled.push(removed);
  }
  return shuffled;
};

const searchInTags = (tags:tagType[], search: string) => {
  for (let i = 0; i < tags.length; i += 1) {
    if (tags[i].tag.toLowerCase().includes(search)) {
      return true;
    }
  }
  return false;
};

export const getFilteredPins = (pinList: PinType[], search: string | null) => {
  if (!search) return pinList;
  const filteredPins = pinList.filter(({ owner, imgDescription, tags }) => {
    const isFoundInDescription = imgDescription.toLowerCase().includes(search);
    const isFoundInOwnerName = owner.name.toLowerCase().includes(search);
    const isFoundInTags = searchInTags(tags, search);
    return isFoundInDescription || isFoundInOwnerName || isFoundInTags;
  });
  return filteredPins;
};

export const getZoomedImageStyle = ({
  naturalWidth: imageWidth,
  naturalHeight: imageHeight,
  getStatic = false,
}:{naturalWidth: number, naturalHeight: number, getStatic?: boolean}) => {
  // dynamically resize image
  let { innerWidth, innerHeight } = window;
  // parameter for innerwidth/height adjustment with mobile consideration
  // top(70) + headingheight(50 / 25) + button height (50 / 25)
  innerHeight = innerHeight < 500 ? innerHeight - 120 : innerHeight - 170;
  // minor x direction adjustment for padding too
  innerWidth -= (innerWidth * 0.02);
  let isNoFit = false;
  if (getStatic) {
    return {
      imgWidth: innerWidth,
      imgHeight: innerHeight,
      parentWidth: innerWidth,
      border: 2,
      isNoFit,
    };
  }
  const ratio = Math.min(innerWidth / imageWidth, innerHeight / imageHeight);
  const newWidth = imageWidth * ratio;
  let parentWidth = newWidth;
  if (newWidth < 500) {
    if (innerWidth < 500) {
      parentWidth = innerWidth;
    } else {
      parentWidth = 500;
    }
    isNoFit = true;
  }
  return {
    imgWidth: newWidth,
    imgHeight: newWidth * (imageHeight / imageWidth),
    parentWidth,
    border: 0,
    isNoFit,
  };
};

export const getFormattedDescription = (imgDescription: string) => (imgDescription.length > 20 ? `${imgDescription.slice(0, 20)}...` : imgDescription);

// eslint-disable-next-line no-promise-executor-return
export const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export const validateURL = (string: string) => {
  try {
    const url = new URL(string);
    if (url.protocol === 'data:' || url.protocol === 'https:') return string;
    if (url.protocol === 'http:') {
      // convert to https to avoid mixed content warning in console
      return `${string.split(':')[0]}s:${string.split(':')[1]}`;
    }
  } catch (_) {
    return null;
  }
  return null;
};

export const getModalWidth = () => {
  const { innerWidth } = window;
  if (innerWidth < 500) {
    return innerWidth;
  } if (innerWidth > 1000) {
    return innerWidth / 3;
  }
  return innerWidth / 2;
};

export const encodeImageFileAsURL = (imgFile: File):Promise<string|ArrayBuffer|null> => new Promise(
  (resolve, reject) => {
    const reader = new FileReader();
    reader.onloadend = () => {
      resolve(reader.result);
    };
    reader.onerror = (err) => {
      reject(err);
    };
    reader.readAsDataURL(imgFile);
  },
);

export const initialDisplayPerScroll = () => (window.innerWidth > 1440 ? 20 : 10);

export const formatDate = (date: string) => {
  const [, day, mth, year] = new Date(date).toUTCString().split(' ');
  return `${day} ${mth} ${year}`;
};
// only good for a single populated pin
export const updatePinList = (oldList: PinType[], newPin: PinType) => {
  const indexOfUpdate = oldList.findIndex((p) => p._id === newPin._id);
  return [
    ...oldList.slice(0, indexOfUpdate),
    newPin,
    ...oldList.slice(indexOfUpdate + 1),
  ];
};

// used for on the fly change of display name in my pins page
export const onTheFlyPinListNameChange = (
  pinList: PinType[],
  newName: string,
  oldDisplayName: string | null,
) => pinList.map((pin) => {
  if (pin.owns) {
    return {
      ...pin,
      owner: {
        ...pin.owner,
        name: newName,
      },
    };
  }
  const newSavedBy = pin.savedBy.map(
    (pinner) => (pinner.name === oldDisplayName
      ? { ...pinner, name: newName }
      : pinner
    ),
  );
  return {
    ...pin,
    savedBy: newSavedBy,
  };
});
