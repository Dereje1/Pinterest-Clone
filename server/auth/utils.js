const getApiKeys = (authType) => {
  const {
    TWITTER_CONSUMER_KEY,
    TWITTER_CONSUMER_SECRET,
    TWITTER_CALLBACK,
    GOOGLE_CLIENT_ID,
    GOOGLE_CLIENT_SECRET,
    GOOGLE_CALLBACK,
  } = process.env;
  if (authType === 'twitter' && TWITTER_CONSUMER_KEY && TWITTER_CONSUMER_SECRET && TWITTER_CALLBACK) {
    return {
      consumerKey: TWITTER_CONSUMER_KEY,
      consumerSecret: TWITTER_CONSUMER_SECRET,
      callbackURL: TWITTER_CALLBACK,
    };
  } if (authType === 'google' && GOOGLE_CLIENT_ID && GOOGLE_CLIENT_SECRET && GOOGLE_CALLBACK) {
    return {
      clientID: GOOGLE_CLIENT_ID,
      clientSecret: GOOGLE_CLIENT_SECRET,
      callbackURL: GOOGLE_CALLBACK,
    };
  }
  console.log(`Missing API keys for --> ${authType}`);
  return null;
};

module.exports = { getApiKeys };
