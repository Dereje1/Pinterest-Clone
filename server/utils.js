const https = require('https');
const Stream = require('stream').Transform;
const AWS = require('aws-sdk');


/* Isolate auth service used from req.user */
const getUserProfile = (user) => {
  const [service] = ['google', 'twitter'].filter(s => user && Boolean(user[s].id));
  return {
    userId: service && user[service].id,
    displayName: service && user[service].displayName,
    service,
  };
};
/* filterPins return only required pin info to the client */
/*
returns
{
    _id, // send full
    imgDescription, // send full
    imgLink, // send full
    owner: name, // send only display name
    savedBy: modifiedSavedBy, // send only display names of pinners
    owns: // need for displaying action button on pin
    hasSaved: // need for displaying action button on pin
    createdDate: send full
};
*/
const filterPins = (rawPins, user) => rawPins.map((pin) => {
  const {
    _id, imgDescription, imgLink, owner, savedBy, createdAt,
  } = pin;
  const { userId } = getUserProfile(user);
  const savedIds = savedBy.map(s => s.id);
  const { name } = owner;
  const modifiedSavedBy = savedBy.map(pinner => pinner.name);
  return {
    _id,
    imgDescription,
    imgLink,
    owner: name,
    savedBy: modifiedSavedBy,
    owns: userId ? userId === owner.id : null,
    hasSaved: userId ? savedIds.includes(userId) : null,
    createdAt,
  };
});

const isReadyToRun = (lastBackedUp) => {
  const CYCLE_TIME = 6 * 60 * 60 * 1000;
  const timeElapsed = new Date() - new Date(lastBackedUp);
  return timeElapsed > CYCLE_TIME;
};

const getPrevBrokenTimeStamp = (prevBrokenPins, pinId) => {
  const [pinFound] = prevBrokenPins.broken.filter(({ _id }) => _id.toString() === pinId.toString());
  if (pinFound) {
    return pinFound.brokenSince;
  }
  return null;
};

const validateURL = (string) => {
  try {
    const url = new URL(string);
    if (url.protocol === 'data:') return 'data protocol';
    if (url.protocol === 'http:' || url.protocol === 'https:') return string;
  } catch (_) {
    return null;
  }
  return null;
};

const isValidEnpoint = imageInfo => new Promise((resolve) => {
  const url = validateURL(imageInfo.imgLink);
  if (url === 'data protocol') {
    resolve({
      ...imageInfo, statusCode: null, statusMessage: 'data protocol', valid: true,
    });
  }
  if (!url) {
    resolve({
      ...imageInfo, statusCode: null, statusMessage: 'Invalid URL', valid: false,
    });
  }

  const inValidStatusMessages = ['Moved Permanently', 'Moved Temporarily'];
  const request = https.request(url, (response) => {
    const { statusCode, statusMessage } = response;
    if (statusCode < 400 && !inValidStatusMessages.includes(statusMessage)) {
      resolve({
        ...imageInfo, statusCode, statusMessage, valid: true,
      });
    } else {
      resolve({
        ...imageInfo, statusCode, statusMessage, valid: false,
      });
    }
  });

  request.on('error', (error) => {
    resolve({
      ...imageInfo, statusCode: null, statusMessage: error, valid: false,
    });
  });

  request.end();
});

const processImage = url => new Promise((resolve, reject) => {
  const urlType = validateURL(url);
  if (!urlType) {
    reject(new Error('Invalid URL type'));
  }

  if (urlType === 'data protocol') {
    const base64Image = url.split(';base64,').pop();
    resolve(Buffer.from(base64Image, 'base64'));
  }

  const request = https.request(url, (response) => {
    const data = new Stream();
    response.on('data', (chunk) => {
      data.push(chunk);
    });

    response.on('end', () => {
      resolve(data.read());
    });
  });

  request.on('error', (err) => {
    reject(err);
  });
  request.end();
});

const configureS3 = () => new AWS.S3({
  accessKeyId: process.env.AWS_ACCESS_KEY_ID,
  secretAccessKey: process.env.AWS_SECRET_KEY,
});

const uploadImageToS3 = async ({ originalImgLink }) => {
  try {
    const s3 = configureS3();
    const Body = await processImage(originalImgLink);
    const Key = `${Date.now()}`;
    const params = {
      Bucket: 'pinterest.clone',
      Key,
      Body,
      ContentType: 'image/png',
    };
    const uploadedImage = await s3.upload(params).promise();
    console.log(`Successfully uploaded image ${originalImgLink} to S3`);
    return uploadedImage.Location;
  } catch (error) {
    console.log(`Error uploading img. ${originalImgLink} - ${error}`);
    return null;
  }
};

module.exports = {
  getUserProfile,
  filterPins,
  isReadyToRun,
  isValidEnpoint,
  getPrevBrokenTimeStamp,
  uploadImageToS3,
  configureS3,
};
