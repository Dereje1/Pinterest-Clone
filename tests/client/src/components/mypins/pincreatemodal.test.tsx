/**
 * @jest-environment jsdom
 */
/* eslint-disable no-import-assign */
import React from 'react';
import { EnzymePropSelector, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import PinCreate from '../../../../../client/src/components/mypins/pincreatemodal';
import * as utils from '../../../../../client/src/utils/utils';

jest.mock('../../../../../client/src/components/mypins/error.png', () => '/load-error.png');
jest.useFakeTimers();

describe('The pin creation modal', () => {
  let props;
  beforeEach(() => {
    props = {
      userInfo: {
        authenticated: true,
        displayName: 'tester displayName',
        username: 'tester username',
        service: 'tester service',
        userId: 'tester user Id',
      },
      message: true,
      allPinLinks: [{ imgLink: 'https://duplicate.com' }],
      reset: jest.fn(),
      savePin: jest.fn(),
    };
  });
  afterEach(() => {
    props = null;
  });

  test('will render for links to images', () => {
    const wrapper = shallow(<PinCreate {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for uploading images', () => {
    const wrapper = shallow(<PinCreate {...props} />);
    const cardHeader: EnzymePropSelector = wrapper.find('ForwardRef(CardHeader)');
    let dialogTitle = wrapper.find('ForwardRef(DialogTitle)');
    expect(dialogTitle.text()).toBe('Create pin from link');
    const uploadSwitch = cardHeader.props().action.props.control.props;
    uploadSwitch.onChange();
    dialogTitle = wrapper.find('ForwardRef(DialogTitle)');
    expect(dialogTitle.text()).toBe('Create pin from file');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will handle errors in images', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    wrapper.setState({ isError: false });
    expect(wrapper.state().isError).toBe(false);
    const img = wrapper.find({ id: 'new-pin-image' });
    img.props().onError();
    expect(wrapper.state().isError).toBe(true);
  });

  test('will handle a loaded image with an error', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    wrapper.setState({ isLoaded: false });
    expect(wrapper.state().isLoaded).toBe(false);
    const img = wrapper.find({ id: 'new-pin-image' });
    img.props().onLoad({ target: { currentSrc: 'http://localhost/load-error.png' } });
    expect(wrapper.state().isLoaded).toBe(true);
    expect(wrapper.state().isError).toBe(true);
  });

  test('will handle a loaded image without an error', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    wrapper.setState({ isLoaded: false });
    expect(wrapper.state().isLoaded).toBe(false);
    const img = wrapper.find({ id: 'new-pin-image' });
    img.props().onLoad({ target: { currentSrc: 'valid' } });
    expect(wrapper.state().isLoaded).toBe(true);
    expect(wrapper.state().isError).toBe(false);
  });

  test('will handle changes in description', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    const description = wrapper.find({ id: 'pin-description' });
    description.props().onChange({ target: { value: 'abc' } });
    expect(wrapper.state().description).toBe('abc');
  });

  test('will handle changes in image links', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    const imgLink = wrapper.find({ id: 'pin-img-link' });
    imgLink.props().onChange({ target: { value: 'https://abc.com' } });
    expect(wrapper.state().picPreview).toBe('https://abc.com');
  });

  test('will handle changes in image links with http', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    const imgLink = wrapper.find({ id: 'pin-img-link' });
    imgLink.props().onChange({ target: { value: 'http://abc.com' } });
    expect(wrapper.state().picPreview).toBe('https://abc.com');
  });

  test('will handle changes in image links with data protocol', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    const imgLink = wrapper.find({ id: 'pin-img-link' });
    imgLink.props().onChange({ target: { value: 'data:abc.com' } });
    expect(wrapper.state().picPreview).toBe('data:abc.com');
  });

  test('will handle changes in image links with invalid url', () => {
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    const imgLink = wrapper.find({ id: 'pin-img-link' });
    imgLink.props().onChange({ target: { value: 'htt://abc.com' } });
    expect(wrapper.state().picPreview).toBe('');
  });

  test('will detect duplicate pins', async () => {
    const wrapper = shallow(<PinCreate {...props} />);
    wrapper.setState({
      picPreview: 'https://duplicate.com',
      description: 'abcde',
      showErrorImage: false,
    });
    const SavePin: EnzymePropSelector = wrapper.find('SavePin');
    expect(SavePin.props().isDuplicateError).toBe(true);
  });

  test('will handle successfully uploading images', async () => {
    jest
      .spyOn(utils, 'encodeImageFileAsURL')
      .mockImplementationOnce(() => Promise.resolve('data:image successfully encoded'));
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    // turn upload switch on
    const cardHeader: EnzymePropSelector = wrapper.find('ForwardRef(CardHeader)');
    const uploadSwitch = cardHeader.props().action.props.control.props;

    wrapper.setState({ isLoaded: true, isError: true });
    uploadSwitch.onChange();
    const instance = wrapper.instance() as PinCreate;
    // TODO: Fix is any declaration
    await instance.handleUploadedImage({ target: { files: ['a file'] } } as any);
    expect(wrapper.state().picPreview).toBe('data:image successfully encoded');
    expect(wrapper.state().isError).toBe(false);
    expect(wrapper.state().isLoaded).toBe(false);
  });

  test('will handle uploading image failures', async () => {
    jest
      .spyOn(utils, 'encodeImageFileAsURL')
      .mockImplementationOnce(() => Promise.reject(new Error()));
    const wrapper = shallow<PinCreate>(<PinCreate {...props} />);
    wrapper.setState({ isLoaded: true, isError: false, picPreview: 'stub_url' });
    const instance = wrapper.instance() as PinCreate;
    // TODO: Fix is any declaration
    await instance.handleUploadedImage({ target: { files: ['a file'] } } as any);
    expect(wrapper.state().picPreview).toBe('');
    expect(wrapper.state().isError).toBe(true);
    expect(wrapper.state().isLoaded).toBe(false);
  });

  test('will save valid pins', async () => {
    const wrapper = shallow(<PinCreate {...props} />);
    wrapper.setState({
      picPreview: 'https://abc.com',
      description: 'abcde',
      showErrorImage: false,
    });
    const SavePin: EnzymePropSelector = wrapper.find('SavePin');
    expect(SavePin.props().isDuplicateError).toBe(false);
    SavePin.props().savePic();
    jest.advanceTimersByTime(500);
    await Promise.resolve();
    expect(props.savePin).toHaveBeenCalledWith({
      imgDescription: 'abcde',
      imgLink: 'https://abc.com',
    });
    expect(wrapper.state()).toMatchObject({
      picPreview: '',
      description: '',
    });
    expect(props.reset).toHaveBeenCalledTimes(1);
  });

  test('will close the dialog', () => {
    const wrapper = shallow(<PinCreate {...props} />);
    const closeButton = wrapper.find({ id: 'close-pin-create-modal' });
    closeButton.props().onClick();
    const color = closeButton.props().sx.color({ palette: { grey: { 50: 'palette color' } } });
    expect(props.reset).toHaveBeenCalledTimes(1);
    expect(color).toBe('palette color');
  });
});
