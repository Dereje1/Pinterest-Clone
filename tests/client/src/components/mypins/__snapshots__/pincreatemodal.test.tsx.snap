// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The pin creation modal will render for links to images 1`] = `
<ForwardRef(Dialog)
  open={true}
  sx={
    {
      "& .MuiDialog-container": {
        "& .MuiPaper-root": {
          "maxWidth": 341.3333333333333,
          "width": "100%",
        },
      },
    }
  }
>
  <ForwardRef(DialogTitle)
    id="pin-create-dialog-title"
  >
    Create pin from link
    <ForwardRef(IconButton)
      aria-label="settings"
      id="close-pin-create-modal"
      onClick={[Function]}
      sx={
        {
          "color": [Function],
          "position": "absolute",
          "right": 8,
          "top": 8,
        }
      }
    >
      <Memo(ForwardRef(CancelIcon))
        style={
          {
            "color": "#3a1c1cde",
            "fontSize": "1.5em",
          }
        }
      />
    </ForwardRef(IconButton)>
  </ForwardRef(DialogTitle)>
  <ForwardRef(Card)
    sx={
      {
        "width": 341.3333333333333,
      }
    }
  >
    <ForwardRef(CardHeader)
      action={
        <ForwardRef(FormControlLabel)
          control={
            <ForwardRef(Switch)
              checked={false}
              onChange={[Function]}
            />
          }
          label={<Memo />}
          sx={
            {
              "color": "900",
              "marginRight": 80,
            }
          }
        />
      }
    />
    <ForwardRef(CardMedia)
      component="img"
      id="new-pin-image"
      image="/load-error.png"
      onError={[Function]}
      onLoad={[Function]}
      sx={
        {
          "height": 240,
          "objectFit": "contain",
        }
      }
    />
    <ForwardRef(CardActions)>
      <div
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "justifyContent": "space-evenly",
            "width": "100%",
          }
        }
      >
        <ForwardRef(TextField)
          color="success"
          error={true}
          id="pin-description"
          label="Description..."
          onChange={[Function]}
          style={
            {
              "margin": "1.5vh",
              "width": "100%",
            }
          }
          value=""
          variant="standard"
        />
        <ForwardRef(TextField)
          color="success"
          error={true}
          id="pin-img-link"
          label="Paste image address here http://..."
          onChange={[Function]}
          style={
            {
              "margin": "1.5vh",
              "width": "100%",
            }
          }
          value=""
          variant="standard"
        />
        <SavePin
          isDescriptionError={true}
          isImageError={true}
          isImageLoaded={false}
          picPreview=""
          savePic={[Function]}
        />
      </div>
    </ForwardRef(CardActions)>
  </ForwardRef(Card)>
</ForwardRef(Dialog)>
`;

exports[`The pin creation modal will render for uploading images 1`] = `
<ForwardRef(Dialog)
  open={true}
  sx={
    {
      "& .MuiDialog-container": {
        "& .MuiPaper-root": {
          "maxWidth": 341.3333333333333,
          "width": "100%",
        },
      },
    }
  }
>
  <ForwardRef(DialogTitle)
    id="pin-create-dialog-title"
  >
    Create pin from file
    <ForwardRef(IconButton)
      aria-label="settings"
      id="close-pin-create-modal"
      onClick={[Function]}
      sx={
        {
          "color": [Function],
          "position": "absolute",
          "right": 8,
          "top": 8,
        }
      }
    >
      <Memo(ForwardRef(CancelIcon))
        style={
          {
            "color": "#3a1c1cde",
            "fontSize": "1.5em",
          }
        }
      />
    </ForwardRef(IconButton)>
  </ForwardRef(DialogTitle)>
  <ForwardRef(Card)
    sx={
      {
        "width": 341.3333333333333,
      }
    }
  >
    <ForwardRef(CardHeader)
      action={
        <ForwardRef(FormControlLabel)
          control={
            <ForwardRef(Switch)
              checked={true}
              onChange={[Function]}
            />
          }
          label={<Memo />}
          sx={
            {
              "color": "900",
              "marginRight": 80,
            }
          }
        />
      }
    />
    <ForwardRef(CardMedia)
      component="img"
      id="new-pin-image"
      image="/load-error.png"
      onError={[Function]}
      onLoad={[Function]}
      sx={
        {
          "height": 240,
          "objectFit": "contain",
        }
      }
    />
    <ForwardRef(CardActions)>
      <div
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "justifyContent": "space-evenly",
            "width": "100%",
          }
        }
      >
        <ForwardRef(TextField)
          color="success"
          error={true}
          id="pin-description"
          label="Description..."
          onChange={[Function]}
          style={
            {
              "margin": "1.5vh",
              "width": "100%",
            }
          }
          value=""
          variant="standard"
        />
        <ForwardRef(Button)
          color="error"
          component="label"
          startIcon={<Memo />}
          sx={
            {
              "margin": "2.3vh",
            }
          }
          variant="contained"
        >
          choose image
          <input
            accept="image/*"
            hidden={true}
            onChange={[Function]}
            type="file"
          />
        </ForwardRef(Button)>
        <SavePin
          isDescriptionError={true}
          isImageError={true}
          isImageLoaded={false}
          picPreview=""
          savePic={[Function]}
        />
      </div>
    </ForwardRef(CardActions)>
  </ForwardRef(Card)>
</ForwardRef(Dialog)>
`;
